# Classification of romantic interests of secondary school students

In [50]:
import sklearn as sk
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

## Importing the data 

In [51]:
df = pd.read_csv('student-por.csv')

In [52]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        4   0  11  11  
1    ...      5        3      3     1     1      3        2   9  11  11  
2    ...      4        3      2     2     3      3        6  12  13  12  
3    ...      3        2      2     1     1      5        0  14  14  14  
4    ...      4        3      2     1     2      5        0  11  13  13  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
644  ...      5        4      2     1     2      5        4  10  11  10  
645  ...      4        3      4     1     1      1        4  15  15  16  
646  ...      1        1      1     1     1      5        6  11  12   9  
647  ...      2        4      5     3     4      2        6  10  10  10  
648  ...      4        4      1     3     4      5        4  10  11  11  

[649 rows x 33 columns]

The data consists of a number of variables. I can't see all the columns, so that's why i use .columns to see all the columnnames. On Kaggle, i saw all the meanings of the variables

In [53]:
df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

Now that I know what variables there are, I will make a selection for my classification. The only problem is that when I chose variables that look interesting, I can't use them because they aren't numerical. That's why I chose the variables that are numerical. I'm not really sure how to fix this! I tried replacing it, but when replace F with 1 and M with 0 (sex) it is only applied to one of the two

In [54]:
df = df[['age', 'famrel', 'freetime', 'goout', 'health', 'romantic', 'absences']]

In [55]:
df

age  famrel  freetime  goout  health romantic  absences
0     18       4         3      4       3       no         4
1     17       5         3      3       3       no         2
2     15       4         3      2       3       no         6
3     15       3         2      2       5      yes         0
4     16       4         3      2       5       no         0
..   ...     ...       ...    ...     ...      ...       ...
644   19       5         4      2       5       no         4
645   18       4         3      4       1       no         4
646   18       1         1      1       5       no         6
647   17       2         4      5       2       no         6
648   18       4         4      1       5       no         4

[649 rows x 7 columns]

## Training the algorithm
I will train the dataset with Random Forest. 

In [56]:
X = df[['age', 'famrel', 'freetime', 'goout', 'health', 'absences']]
y = df['romantic'] 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

In [57]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1, n_estimators=100) 
rf = rf.fit(X_train, y_train)

In [58]:
rf.score(X_train, y_train)

0.973568281938326

I'm not really sure where this score is for?

## Evaluating the model

I will conduct a confusion matrix to look at the accuracy, precision and recall. To see which variable is whcih, I used .classes_. It showed that the first variable is 'no', so I can label my dataframe for the confusion matrix. 

In [59]:
rf.classes_

array(['no', 'yes'], dtype=object)

In [60]:
y_pred = rf.predict(X_test) 
cm = confusion_matrix(y_test, y_pred) 
cm = pd.DataFrame(cm, index=['no (actual)', 'yes (actual)'], columns = ['no (pred)', 'yes (pred)']) 
cm

no (pred)  yes (pred)
no (actual)          99          29
yes (actual)         46          21

In [61]:
21/29

0.7241379310344828

The accuracy is not really high, but looking at the variables I used I wasn't really expecting it to be higher.

In [62]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          no       0.68      0.77      0.73       128
         yes       0.42      0.31      0.36        67

    accuracy                           0.62       195
   macro avg       0.55      0.54      0.54       195
weighted avg       0.59      0.62      0.60       195



The precision and recall for predicting whether the students have a romantic interest is a bit higher than if they do.
It says that it is hard to find out who does have a love interest. The precision says that there is 42% change that the person really has a love interest. If you ask a lot of people, the changes that they actually have a love interest is 31%. That's not very high.

For people without a love interest, the numbers are a little bit higher. If you ask a lot of people, the changes that they actually don't have a love interest is 77%. That's quite high! I don't really know what to say about the precision.

## Conclusion and looking back

I struggled a little bit with this assignment. It didn't work to use the variables I wanted and I am not really sure what all the code really does, like the rf.score (I also couldn't find it anywhere). I was not sure how to interpret and interpret the results